// relationship-type/vocab.js
var getters = require('./getters')

module.exports = {
  id: 'RelationshipType',
  collection: 'relationshipTypes',
  prefixes: require('base').vocab.prefixes,
  description: 'A type of relationship between multiple agents.',
  context: 'hol:RelationshipType',
  properties: {
    id: {
      context: '@id',
      description: 'The identifier of the type of agent relationship.',
      type: 'string',
      format: 'uri'
    },
    description: {
      context: 'schema:description',
      description: 'A description of the type of agent relationship.',
      type: 'string'
    },
    reverse: {
      context: 'hol:reverseRelationshipType',
      description: "The relationship type's associated reverse, if any.",
      $ref: 'RelationshipType',
    },
    labels: {
      context: 'hol:labels',
      description: 'A set of labels for displaying the relationship',
      type: 'string'
    },
    queryLabels: {
      context: 'hol:queryLabels',
      description: 'A set of labels for querying the relationship',
      type: 'string'
    },
    color: {
      context: 'hol:color',
      description: 'The color to represent the relationship.',
      type: 'string'
    }
  }
}
