// state/index.js
import assign from 'lodash.assign'

const data = require('./data.json')


const state = assign(data, {
  relationshipDirections: ['source', 'target', 'context'],
  queries: [],
  queryTypes: [
    {
      id: 'stewards/person/source',
      relationshipType: 'stewards',
      agentType: 'Person',
      direction: 'target',
      labels: ['People who steward {target}']
    },
    {
      id: 'stewards/person/target',
      relationshipType: 'stewards',
      agentType: 'Person',
      direction: 'source',
      labels: ['People stewarded by {source}']
    },
    {
      id: 'stewards/person/context',
      relationshipType: 'stewards',
      agentType: 'Person',
      direction: 'context',
      labels: ['People who steward in {context}']
    },
    {
      id: 'is-member-of/person/target',
      relationshipType: 'is-member-of',
      agentType: 'Person',
      direction: 'target',
      labels: ['People in {target}', 'People who belong to {target}']
    },
    {
      id: 'is-member-of/group/source',
      relationshipType: 'is-member-of',
      agentType: 'Group',
      direction: 'source',
      labels: ['Groups that {source} is a member of']
    },
    {
      id: 'is-member-of/group/target',
      relationshipType: 'is-member-of',
      agentType: 'Group',
      direction: 'target',
      label: [ 'Groups in {target}', 'Groups that are part of {target}']
    },
    {
      id: 'includes/group/target',
      relationshipType: 'includes',
      agentType: 'Group',
      direction: 'target',
      label: [ 'Groups that include {target}' ]
    },
    {
      id: 'mentors/person/source',
      relationshipType: 'mentors',
      agentType: 'Person',
      direction: 'source',
      label: [ 'People mentored by {source}' ]
    }


  ],
  graph: { queries: [] }
})

export default state

import { relationshipsByAgentId, relationshipsByAgentIdAndDirection } from 'relationship/getters'
console.log(relationshipsByAgentId(state)['simon'])
console.log(relationshipsByAgentIdAndDirection(state)['simon'])

import { relationshipTypesByAgentId } from 'relationship-type/getters'
console.log(relationshipTypesByAgentId(state)['simon'])

import { queryTypesByRelationshipTypeId, queryTypesByRelationshipTypeIdAndDirection } from 'query-type/getters'
console.log(queryTypesByRelationshipTypeId(state)['is-member-of'])
console.log(queryTypesByRelationshipTypeIdAndDirection(state)['is-member-of'])
