// state/index.js
const state = {
  agentTypes: [
    {
      id: 'Person'
    },
    {
      id: 'Group'
    }
  ],
  agents: [
    {
      id: 'mikey',
      type: 'Person',
      name: 'Mikey',
      image: 'http://images.clipartpanda.com/t-rex-dinosaur-clip-art-T-Rex-Dinosaur_1.png'
    },
    {
      id: 'simon',
      type: 'Person',
      name: 'Simon',
      image: 'http://www.cats.org.uk/uploads/images/pages/photo_latest14.jpg'
    },
    {
      id: 'holodex',
      type: 'Group',
      name: 'Holodex',
      image: 'https://blog.dinosaur.is/content/images/2014/Oct/nation-holon1.png'
    },
    {
      id: 'enspiral',
      type: 'Group',
      name: 'Enspiral',
      image: 'https://logo.clearbit.com/enspiral.com'
    }
  ],
  relationshipTypes: [
    {
      id: 'is-member-of',
      name: 'is member of',
      labels: ['{source} belongs to {target}', '{source} is part of {target}'],
      color: '#3F51B5'
    },
    {
      id: 'stewards',
      name: 'stewards',
      labels: ['{source} stewards {target}'],
      color: '#F44336'
    }
  ],
  relationships: [
    {
      id: 'mikey-is-member-of-holodex',
      type: 'is-member-of',
      source: 'mikey',
      target: 'holodex',
    },
    {
      id: 'simon-is-member-of-holodex',
      type: 'is-member-of',
      source: 'simon',
      target: 'holodex',
    },
    {
      id: 'holodex-is-member-of-enspiral',
      type: 'is-member-of',
      source: 'holodex',
      target: 'enspiral',
    },
    {
      id: 'mikey-stewards-simon',
      type: 'stewards',
      source: 'mikey',
      target: 'simon',
      context: 'enspiral'
    }
  ]
}

export default state

import { relationshipsByAgentId } from 'relationship/getters'
console.log(relationshipsByAgentId(state)['simon'])
