// state/index.js
const state = {
  agentTypes: [
    {
      id: 'Person'
    },
    {
      id: 'Group'
    }
  ],
  agents: [
    {
      id: 'mikey',
      type: 'Person',
      displayName: 'Mikey',
      image: 'http://images.clipartpanda.com/t-rex-dinosaur-clip-art-T-Rex-Dinosaur_1.png'
    },
    {
      id: 'simon',
      type: 'Person',
      displayName: 'Simon',
      image: 'http://www.cats.org.uk/uploads/images/pages/photo_latest14.jpg'
    },
    {
      id: 'holodex',
      type: 'Group',
      displayName: 'Holodex',
      image: 'https://blog.dinosaur.is/content/images/2014/Oct/nation-holon1.png'
    },
    {
      id: 'enspiral',
      type: 'Group',
      displayName: 'Enspiral',
      image: 'https://logo.clearbit.com/enspiral.com'
    }
  ],
  relationshipDirections: ['source', 'target', 'context'],
  relationshipTypes: [
    {
      id: 'is-member-of',
      displayName: 'is member of',
      labels: ['{source} belongs to {target}', '{source} is part of {target}'],
      color: '#3F51B5'
    },
    {
      id: 'stewards',
      displayName: 'stewards',
      labels: ['{source} stewards {target}'],
      color: '#F44336'
    },
    {
      id: 'includes',
      displayName: 'includes',
      labels: ['{source} includes {target}'],
      color: '#3F51B5'
    }
  ],
  relationships: [
    {
      id: 'mikey-is-member-of-holodex',
      type: 'is-member-of',
      source: 'mikey',
      target: 'holodex',
    },
    {
      id: 'holodex-includes-mikey',
      type: 'includes',
      source: 'holodex',
      target: 'mikey'
    },
    {
      id: 'simon-is-member-of-holodex',
      type: 'is-member-of',
      source: 'simon',
      target: 'holodex',
    },
    {
      id: 'enspiral-includes-holodex',
      type: 'includes',
      source: 'enspiral',
      target: 'holodex'
    },
    {
      id: 'holodex-includes-simon',
      type: 'includes',
      source: 'holodex',
      target: 'simon'
    },
    {
      id: 'holodex-is-member-of-enspiral',
      type: 'is-member-of',
      source: 'holodex',
      target: 'enspiral',
    },
    {
      id: 'mikey-stewards-simon',
      type: 'stewards',
      source: 'mikey',
      target: 'simon',
      context: 'enspiral'
    },
    {
      id: 'simon-stewarded-by-mikey',
      type: 'stewards',
      source: 'simon',
      target: 'mikey',
      context: 'enspiral'
    }
  ],
  queries: [],
  queryTypes: [
    {
      id: 'stewards/person/source',
      relationshipType: 'stewards',
      agentType: 'Person',
      direction: 'target',
      labels: ['People who steward {target}']
    },
    {
      id: 'stewards/person/target',
      relationshipType: 'stewards',
      agentType: 'Person',
      direction: 'source',
      labels: ['People stewarded by {source}']
    },
    {
      id: 'stewards/person/context',
      relationshipType: 'stewards',
      agentType: 'Person',
      direction: 'context',
      labels: ['People who steward in {context}']
    },
    {
      id: 'is-member-of/person/target',
      relationshipType: 'is-member-of',
      agentType: 'Person',
      direction: 'target',
      labels: ['People in {target}', 'People who belong to {target}']
    },
    {
      id: 'is-member-of/group/target',
      relationshipType: 'is-member-of',
      agentType: 'Group',
      direction: 'target',
      label: [ 'Groups in {target}', 'Groups that are part of {target}']
    },
    {
      id: 'includes/group/target',
      relationshipType: 'includes',
      agentType: 'Group',
      direction: 'target',
      label: [ 'Groups that include {target}' ]
    }
  ],
  graph: { queries: [] }
}

export default state

import { relationshipsByAgentId, relationshipsByAgentIdAndDirection } from 'relationship/getters'
console.log(relationshipsByAgentId(state)['simon'])
console.log(relationshipsByAgentIdAndDirection(state)['simon'])

import { relationshipTypesByAgentId } from 'relationship-type/getters'
console.log(relationshipTypesByAgentId(state)['simon'])

import { queryTypesByRelationshipTypeId, queryTypesByRelationshipTypeIdAndDirection } from 'query-type/getters'
console.log(queryTypesByRelationshipTypeId(state)['is-member-of'])
console.log(queryTypesByRelationshipTypeIdAndDirection(state)['is-member-of'])
