// layout/index.js
var React = require('react')
var r = require('r-dom')
var debug = require('debug')('holodex:ui:layout')
var Sidebar = require('sidebar')

var RawView = require('base/views/raw')
var isDev = require('util/isDev')

var style = require('./style')

module.exports = React.createClass({

  getInitialState: function () {
    return {
      leftPane: true,
      rightPane: false
    }
  },

  render: function () {
    debug('render', this.props)
    var self = this
    var leftPaneActive = self.state.leftPane
    var leftWrapperClass = leftPaneActive ? '' : 'toggled'
    var route = this.props.route

    return (
      r.div({ id: 'wrapper', className: leftWrapperClass }, [
        r(Sidebar, { toggle: self.toggleLeftPane, isOpen: leftPaneActive }),
        r.div( { id: 'page-content-wrapper'}, [
          isDev ? (
            r.div({
              className: "debug",
            }, [
              r.header({},
                       "collection: " + route.collection
                       + ", id: " + route.id
                       + ", relation: " + route.relation
                       + ", query: " + JSON.stringify(route.query)
                      ),
            ])
          ) : null,
          r.a({
            href: "https://github.com/craftodex/app",
            style: style.projectSource,
            className: "btn btn-default"
          }, "View Source"),
          r.div({}, this.props.children),
        ])
      ])
    )
  },

  toggleLeftPane: function () {
    var bool = !this.state.leftPane
    this.setState( { leftPane: bool })
  }


})
