// sidebar/tag-select.js
var r = require('r-dom')
var React = require('react')
var debug = require('debug')('tag-select')
var keys = require('lodash.keys')
var Typeahead = require('simple-react-typeahead')

module.exports = React.createClass({
  propTypes: {
    style: React.PropTypes.object,
    tagObject: React.PropTypes.object.isRequired,
    type: React.PropTypes.string.isRequired,
    updateActiveTag: React.PropTypes.func.isRequired
  },

  render: function () {
    var tagObject = this.props.tagObject
    var options = []

    keys(tagObject).forEach(function (key) {
      var tag = tagObject[key]
      if (tag.model && !tag.active ) {
        var option = this.props.type === 'linkLabels' ? key : tag.model.pluralName
        options.push(option)
      }
    }.bind(this))

    return r.div({ className: this.props.type+'-control' }, [
      r(Typeahead, {
          id: 'test',
          style: this.props.style,
          array: options,
          placeholder: 'Enter a ' + this.props.type,
          onSelect: this.handleSelect
        }
      )
    ])
  },

  handleSelect: function (val) {
    debug('handleSelect', val)
    this.props.updateActiveTag(this.props.type, val)
  }
})
