// query-description/reducers.js

// relationship / agentType / relation

const initialState = {
  ids: []
  byId: {
    'stewards/person/source': {
      relationshipType: 'steward',
      agentType: 'Person',
      relation: 'target',
      text: 'People who steward {target}'
    },
    'stewards/person/target': {
      relationshipType: 'steward',
      agentType: 'Person',
      relation: 'source',
      text: 'People stewarded by {source}'
    },
    'stewards/person/context': {
      relationshipType: 'steward',
      agentType: 'Person',
      relation: 'context',
      text: 'People who steward in context'
    },
    'is-member-of/person/target/0': {
      relationshipType: 'is-member-of',
      agentType: 'Person',
      relation: 'target',
      text: 'People who belong to {target}'
    },
    'is-member-of/person/target/1': {
      relationshipType: 'is-member-of',
      agentType: 'Person',
      relation: 'target',
      text: 'People in {target}'
    },
    'is-member-of/group/target/0': {
      relationshipType: 'is-member-of',
      agentType: 'Group',
      relation: 'target',
      text: 'Groups that are part of {target}'
    },
    'is-member-of/group/target/1': {
      relationshipType: 'is-member-of',
      agentType: 'Group',
      relation: 'target',
      text: 'Groups in {target}'
    }
  }
}

module.exports = {
  initialState: initialState
}
