// group/views/page.js
var React = require('react')
var r = require('r-dom')
var debug = require('debug')('craftodex:ui:group:views:page')

var router = require('router')
var GraphView = require('graph-view/graph-view')
var ListView = require('list-view')
var Link = require('base/views/link')

module.exports = React.createClass({
  componentDidMount: function () {
    this.props.model.on('change', function () {
      this.forceUpdate()
    }, this)
  },

  renderList: function () {
    return (
      r(ListView, {
        model: this.props.model
      })
    )
  },

  renderGraph: function () {
    var model = this.props.model
    // TODO actions
    var actions = {}
    debug("model", model)

    var predicates = this.state.predicates.map(function (pred) {
      return this.state.prefix + pred
    }.bind(this))

    debug({predicates: predicates, statePredicates: this.state.predicates })
    return (
      r(GraphView, this.props)
    )
  },

  render: function () {
    var model = this.props.model
    var route = this.props.route

    debug('render', this.props)

    var view
    if (route.query.view === 'list') {
      view = this.renderList()
    } else {
      view = this.renderGraph()
    }

    return (
      r.div({}, [
        r.header({
          className: 'page-header'
        }, [
          r(Link, {
            href: router.format({
              collection: 'groups',
              id: model.id
            })
          }, r.h1({}, model.name)),
        r.header({}, [
        ]),
        view
      ])
    )
  }
})
