// query/vocab.js
var getters = require('./getters')

module.exports = {
  id: 'Query',
  // collection: 'queries',
  prefixes: require('base').vocab.prefixes,
  description: 'A precise requst for information',
  context: 'hol:Query',
  properties: {
    id: {
      context: '@id',
      description: 'The identifier for the query',
      type: 'string',
      format: 'uri'
    },
    relationshipType: {
      context: 'hol:relationshipType',
      description: 'The type of relationship used in the query',
      $ref: 'RelationshipType'
    },
    description: {
      context: 'schema:description',
      description: "a human-readable description of the query",
      type: 'string'
    },
    source: {
      context: 'hol:source',
      description: 'The source agent in the relationship',
      $ref: 'Agent'
    },
    target: {
      context: 'hol:target',
      description: 'The target agent in the relationship',
      $ref: 'Agent'
    },
    context: {
      context: 'hol:context',
      description: 'The context agent in the relationship',
      $ref: 'Agent'
    }
  }
}
