// ui-service/index.js
var Path = require('path')
var Url = require('url')
var fs = require('fs')
var extend = require('xtend')
var express = require('express')
var cookieParser = require('cookie-parser')
var serveStatic = require('serve-static')
var less = require('less-middleware')
var debug = require('debug')('holodex:server')

var serveBundle = require('bundle-service')
var isProd = require('util/isProd')
var isDev = require('util/isDev')

module.exports = uiService

function uiService (config) {

  var app = express()

  //
  // serve less styles
  //
  app.use(less('/', {
    debug: isDev,
    pathRoot: __dirname,
    dest: 'assets',
    once: !isDev,
    parser: {
      relativeUrls: true,
      paths: fs.readdirSync(Path.join(__dirname, '..'))
        .concat([
          Path.join(__dirname, '..', '..', '../node_modules/bootstrap/less')
        ])
    }
  }))

  // serve static files
  app.use(serveStatic(Path.join(__dirname, '..', '..', 'assets')))

  // route to UI
  app.use(function (req, res) {
    var index = Path.join(__dirname, '..', '..', 'index.html')
    res.sendFile(index, function (err) {
      if (err) {
        res.status(err.status).end()
      }
    })
  })

  return app
}
