// graph/action-creators.js
import { createAction } from 'redux-actions'
import {
  layoutUpdating,
  getAgentGraphId,
  getLayout,
  getNodes,
  getIterationsCount } from './getters'

import {
  AGENT_GRAPH,
  GRAPH_START,
  GRAPH_END,
  GRAPH_TICK,
  UPDATE_LAYOUT,
  BEGIN_GRAPH_UPDATE,
  UPDATE_NODES,
  UPDATE_ITERATIONS } from 'action-types'

const debug = require('debug')('graph:action-creators')


export const beginGraphUpdate = createAction(BEGIN_GRAPH_UPDATE)
export const updateNodes = createAction(UPDATE_NODES)
export const updateIterations = createAction(UPDATE_ITERATIONS)

//
// export const loadAgentGraph = function (agentId) {
//   debug('loadAgentGraph', agentId)
//   return {
//     types: [
//       LOAD_AGENT_GRAPH,
//       LOAD_AGENT_GRAPH_SUCCESS,
//       LOAD_AGENT_GRAPH_ERROR
//     ],
//     payload: [
//       agentGraph.bind(null, agentId),
//       graphStart
//     ],
//     sequence: true
//   }
// }

export const loadAgentGraph = function (agentId) {
  debug('loadAgentGraph', agentId)
  return [
    agentGraph(agentId),
    graphStart(),
  ]
}

export const agentGraph = function (agentId) {
  debug('agentGraph', agentId)
  return {
    type: AGENT_GRAPH,
    payload: agentId
  }
}

export const updateLayout = createAction(UPDATE_LAYOUT)
// export const updatingLayout = createAction(UPDATING_LAYOUT)


export const updatingLayout = function (layout) {
  debug('updatingLayout', layout)
  return [
    updateLayout(layout),
    updateNodes(layout.nodes()),
    graphTick()
  ]
}


export const graphStart = function () {
  return (dispatch, getState) => {
    const state = getState()
    if (!layoutUpdating(state)) {
      const layout = getLayout(state)
      debug('graphStart', layout)
      if (layout) {
        let count = 0

        layout.on('tick', () => {
          if (count < 200) {
            dispatch(updateNodes(layout.nodes()))
            count ++
          }
        })

        layout.start(50, 50, 50)
      }
    }
  }
}

// export const graphTick = function () {
//   return (dispatch, getState) => {
//     const state = getState()
//     debug('graphTick', layoutUpdating(state), state)
//     if (layoutUpdating(state)) {
//       const layout = getLayout(state)
//
//     }
//   }
// }
