// link/index.js
var React = require('react')
var r = require('r-dom')
var invariant = require('react/lib/invariant')
var extend = require('xtend')
var debug = require('debug')('holodex:link')

module.exports = function reactLink (topProps) {
  return React.createClass({
    propTypes: {
      href: React.PropTypes.string
    },

    render: function () {
      debug('render', this.props)

      return (
        r.a(extend(
          topProps,
          this.props,
          {
            onClick: this.handleClick
          }
        ), this.props.children)
      )
    },

    href: function () {
      return this.props.href ||
        topProps.href ||
        invariant(
          false,
          'provide "href" prop to Link component'
      )
    },

    navigate: function () {
      return this.props.navigate ||
        topProps.navigate ||
        invariant(
          false,
          'provide "navigate" prop to Link component'
      )
    },

    handleClick: function (e) {
      e.stopPropagation()
      e.preventDefault()

      this.navigate()(this.href())
    }
  })
}
