// tooltip/index.js
var React = require('react')
var r = require('r-dom')
var Card = require('material-ui/lib/card').Card
var CardHeader = require('material-ui/lib/card').CardHeader
var debug = require('debug')('tooltip')

module.exports = React.createClass({

  contextTypes: {
    muiTheme: React.PropTypes.any
  },

  propTypes: {
    data: React.PropTypes.object, // agent model
    top: React.PropTypes.number,
    left: React.PropTypes.number
  },

  render: function () {
    debug('render legend-widget', this.props)

    return (
      r(Card, {
        expandable: false,
        style: {
          position: 'absolute',
          top: this.props.top,
          left: this.props.left,
        }
      },
      [
        r(CardHeader, {
          title: this.props.data.name,
          // titleColor: React.PropTypes.string,
          // titleStyle: React.PropTypes.object,
          // subtitle: this.props.data.handle,
          // subtitleColor: React.PropTypes.string,
          // subtitleStyle: React.PropTypes.object,
          // textStyle: React.PropTypes.object,
          expandable: false,
        })
      ]
    )
  )
  }
})
