// tooltip/index.js
var React = require('react')
var r = require('r-dom')
var Card = require('material-ui/lib/card').Card
var CardHeader = require('material-ui/lib/card').CardHeader
var Avatar = require('material-ui/lib/avatar')

var debug = require('debug')('tooltip')
var m = require('matrix-zoom')
var getCorner = require('util/get-corner')

module.exports = React.createClass({

  contextTypes: {
    muiTheme: React.PropTypes.any
  },

  propTypes: {
    focusAgent: React.PropTypes.object, // agent model
    center: React.PropTypes.array
  },

  getInitialState: function () {
    return { show: false, width: 0, height: 0 }
  },

  componentDidMount: function () {
    var DOMNode = React.findDOMNode(this)
    debug(DOMNode.getBoundingClientRect())
    this.setState({
      width: DOMNode.getBoundingClientRect().width,
      height: DOMNode.getBoundingClientRect().height,
      show: true
    })
  },

  render: function () {
    debug('render tooltip', this.props, this.state)

    var corner = getCorner({
      corners: this.props.focusAgent.corners,
      coords: [this.props.focusAgent.corners.center.x, this.props.focusAgent.corners.center.y],
      width: this.state.width,
      height: this.state.height,
      center: this.props.center,
      screenWidth: this.props.screenWidth,
      screenHeight: this.props.screenHeight
    })

    return (
      r(Card, {
        expandable: false,
        style: {
          transition: 'none',
          opacity: this.state.show ? 1 : 0,
          position: 'absolute',
          top: corner.y,
          left: corner.x
        }
      },
      [
        r(CardHeader, {
          title: this.props.focusAgent.data.name,
          // titleColor: React.PropTypes.string,
          // titleStyle: React.PropTypes.object,
          // subtitle: this.props.data.handle,
          // subtitleColor: React.PropTypes.string,
          // subtitleStyle: React.PropTypes.object,
          // textStyle: React.PropTypes.object,
          expandable: false,
        })
      ]
    )
  )
  },




})
