// tooltip/index.js
var React = require('react')
var r = require('r-dom')
var Card = require('material-ui/lib/card').Card
var CardHeader = require('material-ui/lib/card').CardHeader
var debug = require('debug')('tooltip')
var m = require('matrix-zoom')

module.exports = React.createClass({

  contextTypes: {
    muiTheme: React.PropTypes.any
  },

  propTypes: {
    focusAgent: React.PropTypes.object, // agent model
    center: React.PropTypes.array
  },

  getInitialState: function () {
    return { show: false, width: 0, height: 0 }
  },

  componentDidMount: function () {
    var DOMNode = React.findDOMNode(this)
    debug(DOMNode.getBoundingClientRect())
    this.setState({ width: DOMNode.getBoundingClientRect().width, show: true })
  },

  render: function () {
    debug('render tooltip', this.props, this.state)

    // var top = this.props.focusAgent.coords[1] +
    //   this.props.focusAgent.size / 2
    //
    // var left = this.props.focusAgent.coords[0] +
    //   this.props.focusAgent.size / 2

    var top = this.props.focusAgent.corners.se.y

    var left = this.props.focusAgent.corners.se.x

    // var p = m([this.props.center[0], this.props.  center[1], 1], this.props.vector)
    // var q = [this.props.center[0] - p[0], this.props.center[1] - p[1]]
    // debug(q, this.props.vector)


    // var transform = 'translate(' +
    //   // (p[0] * 2) +
    //   (this.props.vector[0] * 2) + q[0] +
    //   'px,' +
    //   // p[1] * 2 +
    //   this.props.vector[1] * 2 + q[1] +
    //   'px)'

    return (
      r(Card, {
        expandable: false,
        style: {
          // transform: transform,
          transition: 'none',
          opacity: this.state.show ? 1 : 0,
          position: 'absolute',
          top: top,
          left: left,
        }
      },
      [
        r(CardHeader, {
          title: this.props.focusAgent.data.name,
          // titleColor: React.PropTypes.string,
          // titleStyle: React.PropTypes.object,
          // subtitle: this.props.data.handle,
          // subtitleColor: React.PropTypes.string,
          // subtitleStyle: React.PropTypes.object,
          // textStyle: React.PropTypes.object,
          expandable: false,
        })
      ]
    )
  )
  },




})
