// person/views/page.js
  var React = require('react/addons');
var Bootstrap = require('react-bootstrap');
var _ = require('lodash');
var config = require('uiconfig');
var debug = require('debug')("craftodex:ui:person:views:page")

var router = require('router');
var registry = require('registry');
var Group = require('group/model');
var GraphView = require('graph-view/graph-view');
var ListView = require('list-view');
var Link = require('base/views/link');

var labelMap = require('label-map');

module.exports = React.createClass({

  mixins: [],

  propTypes: {
    model: React.PropTypes.object
  },

  componentDidMount: function () {
    this.props.model.on("change", function () {
      this.forceUpdate();
    }, this);
  },

  renderList: function () {
    return (
      <ListView model={this.props.model} />
    );
  },

  // renderGraph: function () {
  //   var model = this.props.model;
  //  // TODO actions
  //   var actions = {};
    
  //   debug("model", model)

  //   var predicates = this.state.predicates.map(function (pred) { 
  //     return this.state.prefix + pred 
  //   }.bind(this));
 
  //   debug({predicates: predicates, statePredicates: this.state.predicates })

  //   return (
  //     <GraphView 
  //       model={model} 
  //       predicates={predicates} 
  //       actions={actions} />
  //   );
  // },

  render: function () {
    var model = this.props.model;
    var route = this.props.route;
    var cx = React.addons.classSet;

    var imgClasses = cx({
      'media': true,
      'profile-avatar-desktop': true
    })


    debug("render", this.props);
    // var view;
    // if (route.query.view === 'list') {
    //   view = this.renderList();
    // } else {
    //   view = this.renderGraph();
    // }

    var view = this.renderList()

    return (
      <div className='profile container'>
        <header className='profile-avatar-container'>
          <img src={model.image} className={imgClasses} />
          <Link href={ router.format({ id: model.id }) }>
            <h1 className='name'>
              { model.name }
            </h1>
          </Link>
        </header>
        { view }
      </div>
    );
  },




  
});
