// person/views/page.js
var React = require('react/addons')
var r = require('r-dom')
var Bootstrap = require('react-bootstrap')
var _ = require('lodash')
var debug = require('debug')("craftodex:ui:person:views:page")

var router = require('router')
var Group = require('group/model')
var GraphView = require('graph-view/graph-view')
var ListView = require('list-view')
var Link = require('base/views/link')

var vars = require('ui/vars')

var style = {

  profile: {
    position: 'relative',
    margin: vars.gutterWidth / 2,
  },

  profileName: {
    marginTop: 5,
    fontFamily: "'Helvetica Neue', Helvetica, Arial, sans-serif",
    fontWeight: 'bold',
  },

  profileAvatarDesktop: {
    background: '#fff',
    border: '5px solid #fff',
    borderRadius: 12,
    boxShadow: '0 1px 1px rgba(136,153,166,0.15)',
    height: 200,
    width: 200,
  },
}

module.exports = React.createClass({

  mixins: [],

  propTypes: {
    model: React.PropTypes.object,
  },

  componentDidMount: function () {
    this.props.model.on("change", function () {
      this.forceUpdate()
    }, this)
  },

  renderList: function () {
    return (
      r(ListView, {
        model: this.props.model,
      })
    )
  },

  render: function () {
    var model = this.props.model
    var route = this.props.route

    debug("render", this.props)

    var view = this.renderList()

    return (
      r.div({
        className: 'profile container',
        style: style.profile,
      }, [
        r.header({
          className: 'profile-avatar-container'
        }, [
          r.img({
            classSet: {
              media: true,
            },
            style: style.profileAvatarDesktop,
            src: model.image,
          },
          r(Link, {
            href: router.format({ id: model.id }),
          }), [
            r.h1({
              className: 'name',
              style: style.profileName,
            }, model.name)
          ])
        ]),
        view
      ])
    )
  },
})
