// role-type/model.js
var format = require('format-text')
// var debug = require('debug')('holodex:role-type:model')

var Model = require('base/model')

module.exports = Model.extend({
  typeName: 'RoleType',
  collectionName: 'rolesTypes',

  props: {
    color: 'string',
    name: 'string',
    pluralName: 'string',
    label: 'string',
    queryLabels: 'string',
    relationshipType: 'link',
    linkType: 'link'
  },

  derived: {
    isSource: {
      cache: true,
      deps: ['relationshipType'],
      fn: function () {
        return this.relationshipType.linkTypes.some(function (linkType) {
          return linkType.source === this.id
        }.bind(this))
      }
    },

    isTarget: {
      cache: true,
      deps: ['relationshipType'],
      fn: function () {
        return this.relationshipType.linkTypes.some(function (linkType) {
          return linkType.target === this.id
        }.bind(this))
      }
    },


    linkTypeId: {
      cache: false,
      fn: function () {
        return this.getId() + '/link'
      }
    }
  },

  generateLabel: function (agentsByRole) {
    return format(this.label, agentsByRole)
  }
})
