// stores/filters-store.js
var Reflux = require('reflux')
var Actions = require('actions')
var vars = require('ui/vars')
var m = require('minimal-immutable')
var each = require('lodash.foreach')
var clone = require('lodash.clone')
var Url = require('url')
var find = require('lodash.find')

var config = require('config')
var prefix = Url.format(config.api)
var Labels = require('graph-view/labels')

//DEFAULTS

var member = prefix + '/roleTypes/member'
var subgroup = prefix + '/roleTypes/subgroup'
var subgroupLink = subgroup + '/link'


var debug  = require('debug')('stores:filters-store')

var filtersStore = Reflux.createStore({

  listenables: Actions,

  init: function () {
    this.modelId = null,
    this.filtersByType = m.fromJS({ targetAgent: [], context:[] })
  },

  //handles logic for updating graph
  onUpdateFilters: function (model) {
    debug('onUpdateFilters', model)
    var candidate = m.fromJS(model.filtersByType)
    if (!m.equals(this.filtersByType, candidate)) { //filtersByType has changed
      if (!this.modelId || this.modelId !== model.getId) {
        this.modelId = model.getId()
        this.model = model
        this.filtersByType = m.fromJS(model.filtersByType)
        this.updateListeners()
      }
    }

  },

  onCancelTag: function (payload) {
    debug('onCanelTag', payload)
    if (payload.type === 'AgentType') {
      this.removeAgentTypeTag(payload)
    }
    else if (payload.type === 'Link') {
      var filtersByType = m.toJS(this.filtersByType)
      debug(filtersByType.targetAgent)
      var filter = find(filtersByType.targetAgent, function (filter) {
        return filter.id === payload.filter.id
      })
      debug('filter', filter)
      if (filter.agentType.active) {
        console.log('filtersStore', payload, filter)
        this.removeLink(payload)
        this.updateListeners({ holdPositions: true })
      }
      else {
        this.removeLink(payload)
        this.updateListeners()
      }
    }
  },

  onAddTag: function (payload) {
    debug('onAddTag', payload)
  },

  removeAgentTypeTag: function (payload) {
    debug('removeAgentTypeTag', payload)
    var filtersByType = m.toJS(this.filtersByType)
    var updatedFilters = filtersByType.targetAgent.map(function (filter) {
      if (filter.agentType.id === payload.filter.agentType.id) {
        var updatedFilter = clone(filter)
        debug('updated filter')
        updatedFilter.agentType.active = false
        return updatedFilter
      }
      else { return filter}
    })
    filtersByType.targetAgent = updatedFilters
    this.filtersByType = m.fromJS(filtersByType)
    this.updateListeners()
  },

  removeLink: function (payload) {
    debug('removeLink', payload)
    var filtersByType = m.toJS(this.filtersByType)
    var updatedFilters = filtersByType.targetAgent.map(function (filter) {
      if (filter.id === payload.filter.id) {
        var updatedFilter = clone(filter)
        debug('removeLnk: updated filter', updatedFilter)
        updatedFilter.linkType.active = false
        return updatedFilter
      }
      else { return filter }
    })
    filtersByType.targetAgent = updatedFilters
    this.filtersByType = m.fromJS(filtersByType)
  },

  getFiltersByType: function () {
    return m.toJS(this.filtersByType)
  },

  updateListeners: function (options) {
    debug('updateListeners')
    this.trigger({
      model: this.model,
      filtersByType: m.toJS(this.filtersByType),
      holdPositions: (options && options.holdPositions)
    })
  },



})

module.exports = filtersStore
