// graph-view/get-tag-data.js
var clone = require('lodash.clone')
var keys = require('lodash.keys')
var isUrl = require('is-url-superb')
var Url = require('url')
var config = require('config')
var prefix = Url.format(config.api)

var colors = require('ui/vars').nodesAndLinks

var Map = {}

var membership = prefix + '/relationshipTypes/membership'
var stewardship = prefix + '/relationshipTypes/stewardship'



var tags = {
  triple: {
      subject: {
        type: 'Role',
        id: prefix + '/roleTypes/member',
        color: '#707070',
        active: true,
        label: 'People in',
        withPredLabel: 'People who',
        isPlural: true,
        index: 0
      },
      predicate: {     //NOTE: DOES NOT OFFICIALLY EXIST
        type: 'Link',
        id: prefix + '/roleTypes/member/link', //???
        relationshipType: prefix + '/relationshipTypes/membership',
        color: '#707070',
        active: false,
        label: 'belongs to',
        pluralLabel: 'belong to',
        index: 1
      },
      object: {
        type: 'Group',
        id: prefix + '/groups/enspiral-craftworks',
        color: '#707070',
        active: true,
        label: 'Enspiral Craftworks',
        index: 2
      }
    },

  filters: [
    {
      type: 'Role',
      id: prefix + '/roleTypes/steward',
      color: '#3572b0',
      active: true,
      label: 'a steward',
    },
    {
      type: 'Link',
      relationshipType: prefix + '/relationshipTypes/stewardship',
      id: prefix + '/roleTypes/steward/link',
      color: '#3572b0',
      active: true,
      label: 'stewards',
    },
    {
      type: 'Role',
      id: prefix + '/roleTypes/stewardee',
      color: '#14892c',
      active: false,
      label: 'a stewardee',
    },
    {
      type: 'Link',
      relationshipType: prefix + '/relationshipTypes/stewardship',
      id: prefix + '/roleTypes/stewardee/link',
      color: '#14892c',
      active: false ,
      label: 'is stewarded by',
    }
  ] ,
}


var debug = require('debug')('graph-view:get-relationship-obj')

module.exports = function (graph) {
  var relationships = {}

//   graph.forEachLink(function (link) {
//     var linkTypeId = link.data.type.getId()
//     var linkTypeName = link.data.type.name
//     debug('link', link)
//     var fromId = link.fromId
//     var toId = link.toId
//
//     if (!relationships[linkTypeId]) relationships[linkTypeId] = {}
//
//     link.data.roles.forEach(function (role) {
//       relationships[linkTypeId]
//       var roleTypeId = role.type.getId()
//       debug('role', role)
//
//     })
//
//
// })

return tags

}
