// graph-view/get-queries.js
var store = require('base/store')
var Labels = require('./labels')
var clone = require('lodash.clone')
var keys = require('lodash.keys')
var isArray = require('lodash.isarray')
var pluck = require('lodash.pluck')
var extend = require('xtend')
var filter = require('lodash.filter')
var find = require('lodash.find')
var dot = require('dot-object')
var obj   = require('lodash.zipobject')
var debug = require('debug')('graph-view:get-queries')
var Url = require('url')
var config = require('config')
var prefix = Url.format(config.api)
var parentGroup = prefix + '/roleTypes/parent-group'

//HACK
function composeQueryLabels (qlbls) {
  var queryLabels = {}

  var qlbls = qlbls.split(',')
  ;['source', 'target', 'context'].forEach(function (nodeText) {
    var qlabels = filter(qlbls, function (label) {
      return label.indexOf(nodeText) !== -1
    })
    if (qlabels) {
      queryLabels[nodeText] = qlabels
    }
  })

  return queryLabels
}

function makeLabels (id, options) {
  var relation = options.source ? 'source' : options.target ? 'target' : 'context'
  var componentLabels = options.component.labels.split(',') //HACK
  var queryLabels = composeQueryLabels(options.component.queryLabels) //HACK
  var labels = []
  debug('makeLabels', relation, options)

  if (options[relation] && queryLabels[relation]) {
    queryLabels[relation].forEach(function (label) {
      var format = {}
      format[relation] = options[relation].name
      labels.push({
        showHalo: options.showHalo,
        queryId: id,
        type: 'QueryLabel',
        label: label.format(format),
        color: options.component.color
      })
    })
  }

  if (options.showLink) {
    componentLabels.forEach(function (label) {
      if (label.indexOf(relation) !== -1) { //HACK
        labels.push({
          queryId: id,
          type: 'LinkLabel',
          label: label.format({
            source: dot.pick('source.name', options),
            target: dot.pick('target.name', options),
            context: dot.pick('context.name', options)
          }),
          color: options.component.color
        })
      }
    })
  }

  return labels
}

function getQuery (options) {
  debug('getQuery', options)
    var id = [
      options.component['@type'] +':' + options.component.id,
      options.source ? 'src:' + options.source.id : null,
      options.target ? 'tgt:' + options.target.id : null,
      options.context ? 'ctx:' + options.context.id : null
    ].join('_')

    return {
      'type': 'Query',
      component: options.component,
      id: id,
      source: options.source,
      target: options.target,
      context: options.context,
      labels: makeLabels(id, options),
      showHalo: options.showHalo,
      showLink: options.showLink,
    }
}

module.exports = function (model, activeQueries) {
  var active = []
  var inactive = []
  var contextRelationshipTypes = {}
  var roleTypes = {}
  var contextLinkTypes = {}
  var linkTypes = {}

  debug('activeQueries', activeQueries)

  model.linkTypesByDirection.asSource.forEach(function (linkType) {
    debug(dot.pick(linkType.id + '.sources.' + model.id, activeQueries))
    if (!linkTypes[linkType.id]) {
      var activeQuery = dot.pick(linkType.id + '.sources.' + model.id, activeQueries)
      if (activeQuery) {
        active.push(getQuery({
          source: model,
          component: linkType,
          showLink: activeQuery.showLink,
          showHalo: activeQuery.showHalo
        }))
      }
      else {
        inactive.push(getQuery({
          source: model,
          component: linkType,
          showLink: true
        }))
      }

      linkTypes[linkType.id] = linkType
    }
  })

  model.linkTypesByDirection.asTarget.forEach(function (linkType) {
    if (!linkTypes[linkType.id]) {
      var activeQuery = dot.pick(linkType.id + '.targets.' + model.id, activeQueries)
      if (activeQuery) {
        active.push(getQuery({
          target: model,
          component: linkType,
          showLink: activeQuery.showLink,
          showHalo: activeQuery.showHalo
        }))
      }
      else {
        inactive.push(getQuery({
          target: model,
          component: linkType,
          showLink: true
        }))
      }

      linkTypes[linkType.id] = linkType
    }
  })
  model.contextRelationships.forEach(function (relationship) {
    if (!contextRelationshipTypes[relationship.type.id]) {
      relationship.type.linkTypes.forEach(function (linkType) {
        if (!contextLinkTypes[linkType.id]) {
          var activeQuery = dot.pick(linkType.id + '.contexts.' + model.id, activeQueries)
          debug('activeQuery', activeQuery)
          if (activeQuery) {
            active.push(getQuery({
              context: model,
              component: linkType,
              showLink: activeQuery.showLink,
              showHalo: activeQuery.showHalo
            }))
          }
          else {
            inactive.push(getQuery({
              context: model,
              component: linkType,
              showLink: true
            }))
          }
          contextLinkTypes[linkType.id] = linkType
        }

      })

      contextRelationshipTypes[relationship.type.getId()] = relationship.type
    }
  })

  debug('inactive', inactive, active, model)

  return { inactive: inactive, active: active }
}
