// graph-view/get-tag-data.js
var clone = require('lodash.clone')
var keys = require('lodash.keys')
var isUrl = require('is-url-superb')
var Url = require('url')
var config = require('config')
var prefix = Url.format(config.api)
var vars = require('ui/vars')
var colors = vars.nodesAndLinks
var keys = require('lodash.keys')

var debug = require('debug')('graph-view:get-relationship-obj')
//
var colors = require('ui/vars').nodesAndLinks

var Map = {}

var membership = prefix + '/relationshipTypes/membership'
var stewardship = prefix + '/relationshipTypes/stewardship'
var steward = prefix + '/roleTypes/steward'
var stewardee = prefix + '/roleTypes/stewardee'
var member = prefix + '/roleTypes/member'
var group = prefix + '/roleTypes/group'
var linkLabels = {}
linkLabels[stewardee] = 'is stewarded by'
linkLabels[steward] = 'is the steward of'

function getDefaultTripleFormat (model) {
  if (model['@type'] === 'Group') {

    var defaultTripleFormat = { roleType: member, expression: 'subject', isPlural: true, index: 0, type: 'Role'  }
    debug('defaultTripleFormat', defaultTripleFormat)
     if (model.childGroups.length === 0) {
       defaultTripleFormat.label = 'People in'
       defaultTripleFormat.withPredLabel = 'People who'
       return defaultTripleFormat //People in [Group]
     }
     else {
       defaultTripleFormat.label = 'Groups in'
       defaultTripleFormat.withPredLabel = 'Groups that'
       return defaultTripleFormat //Subgroups in [Group]

     }
   }
   else {
     return { roleType: group, expression: 'object', type: 'Group', isPlural: true } //[Person]'s groups
   }
}

function formatTriple (model, tripleFormat) {
  var triple = {}
  var otherExpression = tripleFormat.expression === 'subject' ? 'object' : 'subject'

  debug('tripleFormat', tripleFormat)
  triple[tripleFormat.expression] = {
    type: tripleFormat.type,
    id: tripleFormat.roleType,
    color: '#707070',
    active: true,
    label: tripleFormat.label,
    withPredLabel: tripleFormat.withPredLabel,
    index: tripleFormat.index,
    isPlural: tripleFormat.isPlural
  }

  triple[otherExpression] = {
    type: model['@type'],
    id: model.getId(),
    color: '#707070',
    active: true,
    label: model.name,
    index: 2
  }

  triple.predicate = {
    type: 'Link',
    id: prefix + '/roleTypes/member/link', //???
    relationshipType: membership,
    color: '#707070',
    active: false,
    label: 'belongs to',
    pluralLabel: 'belong to',
    index: 1
  }
  return triple
}

function makeFilter (role, type, color) {
  return {
    type: type,
    relationshipType: role.relationship.type.getId(),
    id: type === 'Role' ? role.type.getId() : role.linkTypeId,
    color: color  ,
    active: false,
    label: type === 'Role' ? 'a ' + role.type.name : linkLabels[role.type.getId()]
  }
}

module.exports = function (model) {
  var filters = []
  var roleIdMap = {}
  var contextAgentId = model.getId()
  var defaultTripleFormat = getDefaultTripleFormat(model)
  var triple = formatTriple(model, defaultTripleFormat)

  model.relationshipGraph.forEachLink(function (link) {
    link.data.roles.forEach(function (role) {
      if (roleIdMap[role.type.getId()]) { return } //already dealt with
      if (role.relationship.type.getId() === membership) {
        //TODO membership case ???
        if (defaultTripleFormat.type !== role.type.getId()) {
          //another filter option
          roleIdMap[role.type.getId()] = colors[keys(roleIdMap).length]
        }
      }
      else {
        roleIdMap[role.type.getId()] = colors[keys(roleIdMap).length] //keep track of already assigned and give 'em a color'
        filters.push(makeFilter(role, 'Role', roleIdMap[role.type.getId()]))
        filters.push(makeFilter(role, 'Link', roleIdMap[role.type.getId()]))
      }
    })
  })

  debug('Filters', { triple: triple, filters: filters })
  return { triple: triple, filters: filters }

}
