// graph-view/tag-select.js
var r = require('r-dom')
var React = require('react')
var debug = require('debug')('tag-select')
var each = require('lodash.foreach')
var Typeahead = require('simple-react-typeahead')
var Actions = require('actions')

module.exports = React.createClass({

  propTypes: {
    style: React.PropTypes.object,
    options: React.PropTypes.array
  },

  getInitialState: function () {
    var optionsMap = {}
    each(this.props.options, function (option) {
      optionsMap[option.id] = option
    })
    return {
      optionsMap: optionsMap
    }
  },

  render: function () {
    debug('props, state', this.props, this.state)
    return r.div({ className: 'tag-select' }, [
      r(Typeahead, {
          id: 'test',
          style: this.props.style,
          array: this.props.options,
          placeholder: 'Select filters',
          onSelect: this.handleSelect
        }
      )
    ])
  },

  componentWillReceiveProps: function (nextProps) {
    var optionsMap = {}
    each(nextProps.options, function (option) {
      optionsMap[option.id] = option
    })
    this.setState({ optionsMap: optionsMap })
  },

  handleSelect: function (id) {
    var option = this.state.optionsMap[id]
    Actions.selectLabel(option)
  }
})
