// group/model.js
var Agent = require('agent/model')

var Url = require('url')
var config = require('config')
var prefix = Url.format(config.api)
var parentGroup = prefix + '/rolesByType/parent-group'


module.exports = Agent.extend({
  typeName: 'Group',
  collectionName: 'groups',

  derived: {
    parentGroups: {
      cache: false,
      fn: function () {
        return this.relationships.reduce(function (memo, relationship) {
           if (relationship.rolesByType.subgroup
            && relationship.rolesByType.subgroup.agent
            && relationship.rolesByType.subgroup.agent.getId() === this.getId()) {
             memo.push(relationship.rolesByType['parent group'].agent)
           }
           return memo
          }.bind(this), [])
      }
    },
    childGroups: {
      cache: false,
      fn: function () {
        return this.relationships.reduce(function (memo, relationship) {
           if (relationship.rolesByType['parent group']
            && relationship.rolesByType['parent group'].agent.getId() === this.getId()
            && relationship.rolesByType.subgroup.agent ) { // skip over agent === null error) {
              memo.push(relationship.rolesByType.subgroup.agent)
           }
           return memo
          }.bind(this), [])
      }
    },
    label: {
      cache: false,
      fn: function () {
        return this['@type']
      }
    },
    pluralLabel: {
      fn: function () {
        return 'Groups'
      }
    }

  }
})
