// group/model.js
var Agent = require('agent/model')

var Url = require('url')
var config = require('config')
var prefix = Url.format(config.api)


module.exports = Agent.extend({
  typeName: 'Group',
  collectionName: 'groups',


  derived: {
    parentGroups: {
      cache: false,
      fn: function () {
        return this.relationships.reduce(function (memo, relationship) {
           if (relationship.rolesByType.member
           && relationship.rolesByType.member.agent.getId() === this.getId()) {
             memo.push(relationship.rolesByType.group.agent)
           }
           return memo
          }.bind(this), [])
      }
    },
    childGroups: {
      cache: false,
      fn: function () {
        return this.relationships.reduce(function (memo, relationship) {
           if (relationship.rolesByType.group
            && relationship.rolesByType.group.agent.getId() === this.getId()
            && relationship.rolesByType.member.agent // skip over agent === null error
            && relationship.rolesByType.member.agent['@type'] === 'Group') {
              memo.push(relationship.rolesByType.member.agent)
           }
           return memo
          }.bind(this), [])
      }
    },
    label: {
      cache: false,
      fn: function () {
        return this['@type']
      }
    },
    pluralLabel: {
      fn: function () {
        return 'Groups'
      }
    }

  }
})
