// ui/index.js
var React = require('react');

var router = require('router');
var collections = require('collections');

var FourOhFourPage = require('./404');

var GraphView = require('graph-view/graph-view');
var predicates = ['stewardOf'];
//TODO actions
var actions = []

var entities = [
  {
    id: 1,
    name: 'Mikey',
    handle: 'ahdinosaur',
    bio: "i'm a hippie hacker on a mission to demystify the magic of technology and put the people in control. i'm currently working on open source socioeconomic infrastructure within the Enspiral, Open App Ecosystem, and Bevry networks.",
    location: [
      {
        "@id": "http://sws.geonames.org/5332921/",
        "@type": "Place",
        name: "California"
      },
      {
        "@id": "http://sws.geonames.org/2186224/",
        "@type": "Place",
        name: "New Zealand"
      }
    ],
    url: "http://dinosaur.is",
    image: "http://gravatar.com/avatar/22ee24b84d0a2a9446fc9c0fe0652c46?s=512&d=identicon",
    memberships: [ { "@id": "http://api.craftodex.enspiral.info/memberships/craftworks-mikey" } ],
    groups: [ { "@id": "http://api.craftodex.enspiral.info/memberships/craftworks-mikey" }],
    stewardOf: [2,3]
  },
  {
    id: 2,
    name: 'Hannah',
    image: 'http://gravatar.com/avatar/afa91980b06fae820c502ed90f8431fd?s=512&d=identicon',
    stewardOf: null
  },
  {
    id: 3,
    name: 'Craig',
    image: "http://gravatar.com/avatar/8c1b99675b4ba19460bfed9a04739fc4?s=512&d=identicon",
    stewardOf: [4,5]
  },
  {
    id: 4,
    name: 'Nanz',
    image: "https://loomio-attachments.s3.amazonaws.com/users/uploaded_avatars/000/002/685/large/photo.jpeg?1417128452",
    stewardOf: null
  },
  {
    id: 5,
    name: 'Jesse',
    image: "http://gravatar.com/avatar/0b1941ab5c9e58601c0ada31f5b4b68b?s=512&d=identicon",
    stewardOf: null
  }
];

module.exports = React.createClass({

  render: function () {
    var route = this.props.route;

    var Page;
    if (route) {
      Page = collections[route.collection].Page;
    } else {
      Page = FourOhFourPage;
    }

    return (
      <div>
        <header>
          {
            "collection: " + route.collection
            + ", id: " + route.id
            + ", relation: " + route.relation
            + ", query: " + JSON.stringify(route.query)
          }
        </header>
        <main>
          <Page
            route={this.props.route}
            model={this.props.model} />
          <GraphView entities={entities} predicates={predicates} actions={actions} />
        </main>
      </div>
    );
  },
});
