// legend-item/index.js
var React = require('react')
var r = require('r-dom')
var mui = require('material-ui')
var Toggle = require('material-ui/lib/toggle')
var debug = require('debug')('legend-item')
var SvgIcon = require('material-ui/lib/svg-icon')
var arrowRightPath = 'M23.143 11.96q0 0.188-0.134 0.321l-5.143 4.741q-0.214 0.188-0.469 0.080-0.254-0.121-0.254-0.388v-3h-16.714q-0.188 0-0.308-0.121t-0.121-0.308v-2.571q0-0.188 0.121-0.308t0.308-0.121h16.714v-3q0-0.281 0.254-0.388t0.469 0.067l5.143 4.688q0.134 0.134 0.134 0.308z'
var format = require('format-text')

module.exports = React.createClass({

  propTypes: {
    color: React.PropTypes.string,
    queryId: React.PropTypes.oneOfType([
      React.PropTypes.string,
      React.PropTypes.number
    ]),
    label: React.PropTypes.string,
    hasToggle: React.PropTypes.bool,
    onToggle: React.PropTypes.func,
    showLink: React.PropTypes.bool
  },

  contextTypes: {
    muiTheme: React.PropTypes.any
  },

  shouldComponentUpdate: function () {
    return false
  },

  render: function () {
    debug('render legend-widget', this.props, this.context)
    var label = this.props.label
    label = format(label, { source: '', target: '' })
    var theme = this.context.muiTheme || {
      spacing: {
        desktopGutterMini: 8,
        desktopGutterLess: 16,
        desktopGutter: 24
      }
    }

    return r.div(
      {
        className: 'legend-item',
        style: {
          display: 'flex',
          marginTop: theme.spacing.desktopGutterMini,
          marginBottom: theme.spacing.desktopGutterMini,
          marginLeft: theme.spacing.desktopGutterLess ,
          marginRight: theme.spacing.desktopGutter ,
          flexDirection: 'row'
        }
      },
      [
        r(SvgIcon, {
            color: this.props.color,
            style: {
              marginRight: theme.spacing.desktopGutterMini
            }
          },
          [
            r.path({ d: arrowRightPath })
          ]
        ),
        r(Toggle, {
          ref: 'muiToggle',
          defaultToggled: this.props.showLink,
          onToggle: this.handleToggle(this.props.query),
          label: label,
          labelPosition: 'left',
          labelStyle: {
            fontFamily: 'inherit',
            fontSize: 13,
            textAlign: 'center',
            color: this.props.color
          }
        })
      ]
    )
  },

  handleToggle: function (queryId) {
    return function (e, toggled) {
      debug('handleToggle', e, toggled, queryId)
      this.props.onToggle({ queryId: queryId, showLink: toggled })
    }.bind(this)
  }
})
