// list-view/index.js
var React = require('react')
var Bootstrap = require('react-bootstrap')
var r = require('r-dom')
var _ = require('lodash')
var debug = require('debug')('holodex:ui:list-view')

var Link = require('base/views/link')
var registry = require('registry')
var router = require('router')

var style = require('./style')

module.exports = React.createClass({

  render: function () {
  
    var model = this.props.model;

    debug('model', model)

    var content = _.map(model.relsByType, function (rels, relTypeId) {

      var relType = registry.get(relTypeId)

      debug('relType', relType, rels)

      return (
        r.section({}, [
          r.header(relType.label + ':'),

          r(Bootstrap.ListGroup, {}, rels.map(function (rel) {
            debug("rel", rel)

            return (
              r(Bootstrap.ListGroupItem, {
                style: style.listItem,
              }, [
                r(Link, {
                  href: router.format({ id: rel.has.id }),
                }, [
                  r.img({
                    className: "media",
                    style: style.listItemImage,
                    src: rel.has.image,
                  }),
                  r.span({
                    style: style.listItemText,
                  }, rel.has.name),
                ])
              ])
            )
          }))
        ])
      )
    })

    return r.div({}, content)
  },
})
