// relationship/vocab.js
var getters = require('./getters')

module.exports = {
  id: 'Relationship',
  collection: 'relationships',
  prefixes: require('base').vocab.prefixes,
  description: 'A relationship between multiple agents.',
  context: 'hol:Relationship',
  properties: {
    id: {
      context: '@id',
      description: 'The identifier for the relationship',
      type: 'string',
      format: 'uri'
    },
    relationshipType: {
      context: 'hol:relationshipType',
      description: 'The type of agent relationship.',
      $ref: 'RelationshipType'
    },
    description: {
      context: 'schema:description',
      description: "A description of the agents' relations.",
      type: 'string'
    },
    reverse: {
      context: 'hol:reverseRelationship',
      description: "The relationship's associated reverse, if any.",
      $ref: 'Relationship',
    },
    source: {
      context: 'hol:source',
      description: 'The source agent in the relationship',
      $ref: 'Agent'
    },
    target: {
      context: 'hol:target',
      description: 'The target agent in the relationship',
      $ref: 'Agent'
    },
  }
}
