// api/collections.js
// describe collections
var collections = {
  people: require('person/vocab'),
  groups: require('group/vocab'),
  relationshipTypes: require('relationship-type/vocab'),
  relationships: require('relationship/vocab'),
};

// fill in any necessary defaults
var BaseService = require('base-service');
var FsStorage = require('fs-storage');

Object.keys(collections).forEach(function (collName) {

  var collection = collections[collName];

  if (!collection.service) {
    // create default service from base service
    collection.service = new BaseService({
      type: collection.id,
      collection: collName,
      context: {},
      storage: new FsStorage({
        collection: collName,
        collections: collections,
        serialize: function (obj) {
          var yaml = require('js-yaml');
          return yaml.safeDump(obj);
        },
        deserialize: function (str) {
          var yaml = require('js-yaml');
          return yaml.safeLoad(str)
        },
        extname: ".yml",
      })
    });
  }
});

module.exports = collections;
